src/coreclr/src/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs

```diff
@@ -3957,112 +3951,196 @@ private void CreateInstanceCheckThis()
+                        }
+                        _pfnCtorStub = &ValueTypeUnboxingStub;
+                        _ctorStubState = RuntimeMethodHandle.GetFunctionPointer(defaultCtorRMH);
+                    }
+                    else
+                    {
+                        // ValueType with no explicit parameterless ctor; assume ctor returns default(T)
+
+                        _pfnCtorStub = &CtorNoopStub;
+                        _ctorStubState = IntPtr.Zero;
+                    }
                 }
-                _isFullyInitialized = true;
-            }
+                else
+                {
+                    // Reference type - we can't proceed unless there's a default ctor we can call.
 
-            public void EnsureInitialized()
-            {
-                if (!_isFullyInitialized)
-                    Initialize();
-            }
-        }
+                    Debug.Assert(rt.IsClass);
 
-        /// <summary>
-        /// The slow path of CreateInstanceDefaultCtor
-        /// </summary>
-        private object? CreateInstanceDefaultCtorSlow(bool publicOnly, bool wrapExceptions, bool fillCache)
-        {
-            RuntimeMethodHandleInternal runtimeCtor = default;
-            bool canBeCached = false;
-            bool hasNoDefaultCtor = false;
+                    if (_pMT->IsComObject)
+                    {
+                        // COM RCW types go through the runtime's special allocator, but no ctor
+                        // ever gets run over them. We won't overwrite the _pfnCtorStub field.
 
-            object instance = RuntimeTypeHandle.CreateInstance(this, publicOnly, wrapExceptions, ref canBeCached, ref runtimeCtor, ref hasNoDefaultCtor);
-            if (hasNoDefaultCtor)
-            {
-                throw new MissingMethodException(SR.Format(SR.Arg_NoDefCTor, this));
+                        Debug.Assert(!rt.IsGenericCOMObjectImpl(), "__ComObject base class should've been blocked.");
+                        defaultCtorRMH = default; // ignore any parameterless ctor
+                    }
+                    else
+                    {
+                        if (!_pMT->HasDefaultConstructor)
+                        {
+                            throw new MissingMethodException(SR.Format(SR.Arg_NoDefCTor, rt));
+                        }
+
+                        // Reference type with explicit parameterless ctor typed as (object) -> void.
+                        // We'll pass the actual ctor address as the state object.
+
+                        static void ReferenceTypeStub(object? @this, IntPtr state)
+                        {
+                            ((delegate*<object, void>)state)(@this!);
+                        }
+                        _pfnCtorStub = &ReferenceTypeStub;
+                        _ctorStubState = RuntimeMethodHandle.GetFunctionPointer(defaultCtorRMH);
+                    }
+                }
+
+                if (defaultCtorRMH.IsNullHandle())
```

Comment:
can `defaultCtorRMH` be null here? For the COM path above.

Reply:
It'll be null in the case of COM, struct with no parameterless ctor (not an error), and ref type with no parameterless ctor (error). I'll update the asserts to make this clearer.